**NumPy**

- `Num -> Numerical Py -> Python`
- `NumPy -> Numerical Python => we are going to use to work with data which is in the form of dimension `

In [ ]:
# If we want to work with data which is in dimension or
# If we want to create data in the form of dimension we need to use NumPy

In [ ]:
import numpy as np

In [ ]:
len(dir(np))

532

**out of 532 Built_in_functions we are going to learn only the Built_in_functions which are used in AI (DA and DS)**





- `array `
- `arange `
- `ndim `
- `shape `
- `size`
- `reshape `
- `reshape (-1,1) and (1,-1)`
- `zeros `
- `ones `
- `full `
- `eye `
- `diagonal `
- `ravel `
- `transpose `
- `concat`
- `random -> randint `
- `random -> randn`
- `random -> rand`
- `random -> random `
- `max `
- `sum `
- `min `
- `argmax `
- `argmin  `
- `repeat `
- `axis =1`
- `axis = 0 `
- `add and math function`
- `matmul `


In [ ]:
# array = which is used to create data in the form of dimension

f = np.array([10,14,24,35,60]) # 1d

print(f) # calling from Memory

print(f.ndim) # ndim (number of dimensions)

print(f.shape) # shape(rows and columns)

[10 14 24 35 60]
1
(5,)


In [ ]:
print(f[2])

24


In [ ]:
t = []

for i in range(11):
  t.append(i)

y = np.array(t)

print(y)
print(y.ndim)
print(y.shape)

[ 0  1  2  3  4  5  6  7  8  9 10]
1
(11,)


In [ ]:
print(y[6])

6


In [ ]:
# 2d

a = np.array([  [10,1,9]  , [3,6,7] , [2,4,8]  ])

print(a)
print(a.ndim)
print(a.shape)
print(a.size) # (total values)

[[10  1  9]
 [ 3  6  7]
 [ 2  4  8]]
2
(3, 3)
9


In [ ]:
print(a[1][2])

7


In [ ]:
# changing 1D to 2D
a = np.array([10,11,14,17,19,16])
print(a)
print(a.ndim)
print(a.shape)

b = a.reshape(2,3)
print(b)
print(b.ndim)
print(b.shape)

[10 11 14 17 19 16]
1
(6,)
[[10 11 14]
 [17 19 16]]
2
(2, 3)


In [ ]:
# arange = only 1D data with (n-1) logic

a = np.arange(10,21)
print(a)
print(a.ndim)

[10 11 12 13 14 15 16 17 18 19 20]
1


In [ ]:
# 1D to 2D

a = np.arange(1,61)
print(a)
print(a.ndim)
print(a.shape)

b = a.reshape(10,6)
print(b)
print(b.shape)
print(b.ndim)

[ 1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24
 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48
 49 50 51 52 53 54 55 56 57 58 59 60]
1
(60,)
[[ 1  2  3  4  5  6]
 [ 7  8  9 10 11 12]
 [13 14 15 16 17 18]
 [19 20 21 22 23 24]
 [25 26 27 28 29 30]
 [31 32 33 34 35 36]
 [37 38 39 40 41 42]
 [43 44 45 46 47 48]
 [49 50 51 52 53 54]
 [55 56 57 58 59 60]]
(10, 6)
2


In [ ]:
# 1D to 2D

a = np.arange(1,61).reshape(10,6)
print(a)
print(a.ndim)
print(a.shape)


[[ 1  2  3  4  5  6]
 [ 7  8  9 10 11 12]
 [13 14 15 16 17 18]
 [19 20 21 22 23 24]
 [25 26 27 28 29 30]
 [31 32 33 34 35 36]
 [37 38 39 40 41 42]
 [43 44 45 46 47 48]
 [49 50 51 52 53 54]
 [55 56 57 58 59 60]]
2
(10, 6)


In [ ]:
print(a[2][1])

14


In [ ]:
# 1D to 3D

a = np.arange(1,61).reshape(2,5,6)
print(a)
print(a.ndim)
print(a.shape)


[[[ 1  2  3  4  5  6]
  [ 7  8  9 10 11 12]
  [13 14 15 16 17 18]
  [19 20 21 22 23 24]
  [25 26 27 28 29 30]]

 [[31 32 33 34 35 36]
  [37 38 39 40 41 42]
  [43 44 45 46 47 48]
  [49 50 51 52 53 54]
  [55 56 57 58 59 60]]]
3
(2, 5, 6)


In [ ]:
print(a[1][2][3])

46


In [ ]:
print(a[3][0][4])

IndexError: index 3 is out of bounds for axis 0 with size 2

In [ ]:
# reshape(-1,1) and reshape(1,-1)
# 1D to 2D

In [ ]:
a = np.array([10,11,14,7,9])
print(a)
print(a.ndim)
print(a.shape)

b = a.reshape(-1,1)
print(b)
print(b.shape)
print(b.ndim)

In [ ]:
a = np.array([10,11,14,7,9])
print(a)
print(a.ndim)
print(a.shape)

b = a.reshape(1,-1)
print(b)
print(b.shape)
print(b.ndim)

[10 11 14  7  9]
1
(5,)
[[10 11 14  7  9]]
(1, 5)
2


In [ ]:
# 2d to 3D

a = np.array([  [10,1,9]  , [3,6,7] , [2,4,8]  ])
print(a)
print(a.ndim)
print(a.shape)


b = a.reshape(1,3,3)
print(b)
print(b.ndim)
print(b.shape)

[[10  1  9]
 [ 3  6  7]
 [ 2  4  8]]
2
(3, 3)
[[[10  1  9]
  [ 3  6  7]
  [ 2  4  8]]]
3
(1, 3, 3)


In [ ]:
# zeros

a = np.zeros((5,4) , dtype = np.int32)

print(a)

[[0 0 0 0]
 [0 0 0 0]
 [0 0 0 0]
 [0 0 0 0]
 [0 0 0 0]]


In [ ]:
a.dtype

dtype('float64')

In [ ]:
a= np.ones((5,6))

print(a)

[[1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1.]]


In [ ]:
a = np.full((5,4) , 7)
print(a)

[[7 7 7 7]
 [7 7 7 7]
 [7 7 7 7]
 [7 7 7 7]
 [7 7 7 7]]


In [ ]:
a = np.eye(3)

print(a)

[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


In [ ]:
a = np.diag([10,20,30,40])
print(a)

[[10  0  0  0]
 [ 0 20  0  0]
 [ 0  0 30  0]
 [ 0  0  0 40]]


In [ ]:
# numpy -> random -> randint

a = np.random.randint(100,500,3)
print(a)

[316 195 413]


In [ ]:
# numpy -> random -> rand
# rand -> random values between 0 to 1

a = np.random.rand(5,6)
print(a)

[[0.80344593 0.66371373 0.44347709 0.40128013 0.69061828 0.95471072]
 [0.28314597 0.16203221 0.62216078 0.08611438 0.50176988 0.95564912]
 [0.06031071 0.23772531 0.17352986 0.59062879 0.43015225 0.22137223]
 [0.7521228  0.08444231 0.57168463 0.03589542 0.30152193 0.52586337]
 [0.91371015 0.44520655 0.81728733 0.0847659  0.32658316 0.23508607]]


In [ ]:
# numpy -> random -> random
# random -> random values between 0 to 1

a = np.random.random((5,6))
print(a)

[[0.85077797 0.10766853 0.21584486 0.50033337 0.17529563 0.37346102]
 [0.02802049 0.76109071 0.33739982 0.90499742 0.57410223 0.82380742]
 [0.42128239 0.82272732 0.19127205 0.64786666 0.62070583 0.88145066]
 [0.36152123 0.42355152 0.62840508 0.56742927 0.60243348 0.11004348]
 [0.06680429 0.6785899  0.85014401 0.75790624 0.73031477 0.55344287]]


In [ ]:
# numpy -> random -> randn
# randn -> random values between -1 ... to 1

a = np.random.randn(5,6)
print(a)

[[-2.26574933 -0.14366055 -1.36877983 -0.61929137 -0.06956877 -1.63254319]
 [-0.90587246  1.29485661 -0.13679513 -0.45308533 -0.25960897 -2.02035212]
 [-1.18447421 -0.9064948   0.05839698 -0.92548775  0.38824513  0.86128253]
 [-0.19660066 -0.93353723 -0.12717482  0.03282942  0.68421286  0.37973458]
 [ 1.13000901 -0.75962449 -1.1119692   1.78630916  0.21383495 -1.24860093]]


In [ ]:
import pandas as pd
a = np.random.random((100,40))
f = pd.DataFrame(a)
f.to_csv('test_data.csv',index=False)

In [ ]:
# ravel = will convert any dimension data into 1D

a = np.array([  [10,1,9]  , [3,6,7] , [2,4,8]  ])
print(a)
print(a.ndim)
print(a.shape)

b = np.ravel(a)
print(b)
print(b.ndim)
print(b.shape)

[[10  1  9]
 [ 3  6  7]
 [ 2  4  8]]
2
(3, 3)
[10  1  9  3  6  7  2  4  8]
1
(9,)


In [ ]:
# transpose = convert rows to columns and columns to rows

a = np.array([  [10,1,9]  , [3,6,7] , [2,4,8]  ])
print(a)

print('================')

b = np.transpose(a)
print(b)

[[10  1  9]
 [ 3  6  7]
 [ 2  4  8]]
[[10  3  2]
 [ 1  6  4]
 [ 9  7  8]]


In [ ]:
# concatination

a = np.array([ [10,1,9],[ 3  ,6  ,7],[ 2 , 4 , 8] ])
b = np.array([ [10,3 ,2] , [ 1 , 6 , 4] , [ 9  ,7 , 8] ])

print(a)
print(b)
print('================')

c = np.concat((a,b))
print(c)

[[10  1  9]
 [ 3  6  7]
 [ 2  4  8]]
[[10  3  2]
 [ 1  6  4]
 [ 9  7  8]]
[[10  1  9]
 [ 3  6  7]
 [ 2  4  8]
 [10  3  2]
 [ 1  6  4]
 [ 9  7  8]]


In [ ]:
# axis = 1 -> row
# axis = 0 -> column

In [ ]:
b = np.array([ [10,3 ,2] , [ 1 , 6 , 4] , [ 9  ,7 , 8] ])

print(np.sum(b))

50


In [ ]:
b = np.array([ [10,3 ,2] , [ 1 , 6 , 4] , [ 9  ,7 , 8] ])

print(np.sum(b , axis=0))

[20 16 14]


In [ ]:
b = np.array([ [10,3 ,2] , [ 1 , 6 , 4] , [ 9  ,7 , 8] ])

print(np.max(b , axis=0))

[10  7  8]


In [ ]:
b = np.array([ [10,3 ,2] , [ 1 , 6 , 4] , [ 9  ,7 , 8] ])

print(np.min(b , axis=1))

[2 1 7]


In [ ]:
b = np.array([ [10,3 ,2] , [ 1 , 6 , 4] , [ 9  ,7 , 8] ])

print(np.argmax(b))

0


In [ ]:
b = np.array([ [10,3 ,2] , [ 1 , 6 , 4] , [ 9  ,7 , 8] ])

print(np.argmin(b))

3


In [ ]:
b = np.array([ [10,3 ,2] , [ 1 , 6 , 4] , [ 9  ,7 , 8] ])

print(np.max(b,axis=1))

[10  6  9]


In [ ]:
b = np.array([ [10,3 ,2] , [ 1 , 6 , 4] , [ 9  ,7 , 8] ])

print(np.argmax(b,axis=1))

[0 1 0]


In [ ]:
b = np.array([ [10,3 ,2] , [ 1 , 6 , 4] , [ 9  ,7 , 8] ])

print(np.argmin(b,axis=1))

[2 0 1]


In [ ]:
# repeat

a = np.arange(10,21)
b = a.reshape(1,-1)


c = np.repeat(b , 5 , axis=0)
print(c)

[[10 11 12 13 14 15 16 17 18 19 20]
 [10 11 12 13 14 15 16 17 18 19 20]
 [10 11 12 13 14 15 16 17 18 19 20]
 [10 11 12 13 14 15 16 17 18 19 20]
 [10 11 12 13 14 15 16 17 18 19 20]]


In [ ]:
# some math operations

a = np.array([ [10,1,9],[ 3  ,6  ,7],[ 2 , 4 , 8] ])
b = np.array([ [10,3 ,2] , [ 1 , 6 , 4] , [ 9  ,7 , 8] ])

print(np.add(a,b))

[[20  4 11]
 [ 4 12 11]
 [11 11 16]]


In [ ]:
# some math operations

a = np.array([ [10,1,9],[ 3  ,6  ,7],[ 2 , 4 , 8] ])
b = np.array([ [10,3 ,2] , [ 1 , 6 , 4] , [ 9  ,7 , 8] ])

print(np.subtract(a,b))

[[ 0 -2  7]
 [ 2  0  3]
 [-7 -3  0]]


In [ ]:
# some math operations

a = np.array([ [10,1,9],[ 3  ,6  ,7],[ 2 , 4 , 8] ])
b = np.array([ [10,3 ,2] , [ 1 , 6 , 4] , [ 9  ,7 , 8] ])

print(np.multiply(a,b))

[[100   3  18]
 [  3  36  28]
 [ 18  28  64]]


In [ ]:
# matmul

a = np.array([ [1,4] , [7,6] ])
b = np.array([ [4,3,2] , [1,8,6] ])

print(np.matmul(a,b))

[[ 8 35 26]
 [34 69 50]]


In [ ]:
# matmul

a = np.array([ [1,4] , [7,6] ])
b = np.array([ [4,3,2] , [1,8,6] , [10,20,30]])

print(np.matmul(a,b))

ValueError: matmul: Input operand 1 has a mismatch in its core dimension 0, with gufunc signature (n?,k),(k,m?)->(n?,m?) (size 3 is different from 2)

In [ ]:
a = np.array([ [10,1,9],[ 3  ,6  ,7],[ 2 , 4 , 8] ])


for i in a:
  print(i)
  print('GM')

[10  1  9]
GM
[3 6 7]
GM
[2 4 8]
GM


In [ ]:
a = np.array([ [10,1,9],[ 3  ,6  ,7],[ 2 , 4 , 8] ])


for i in a:
  for j in i:
    if j % 2 != 0:
      print(j)

1
9
3
7


In [ ]:
# Questions on NumPy

In [ ]:
import numpy as np
import sys
a = np.array([ [1,7,9,10] , [2,4,8,20] , [3,6,5,30] , [1,2,3,4] ])
print(a.shape)
f = 0

try:
  if a.shape[0] == a.shape[1]:
    s = -1
    for i in a:
      f = f + i[s]
      s = s - 1
    print(f)
  else:
    print(f'Diagonal Adding Not possible')

except Exception as e:
  er_ty,er_msg,er_li = sys.exc_info()
  print(f'Error in Line no: {er_li.tb_lineno} -> Due to : {er_ty}')

(4, 4)
25


In [ ]:
a[0]

array([ 1,  7,  9, 10])

In [ ]:
a

array([[ 1,  7,  9, 10],
       [ 2,  4,  8, 20],
       [ 3,  6,  5, 30],
       [ 1,  2,  3,  4]])

In [ ]:
len(a)

4

In [ ]:
t = []
s = 0
for i in range(len(a)):
  for j in a:
    s = s + j[i]

print(s)

115
